Fix concurrency bug in `NamespacedHierarchicalStore.computeIfAbsent`

martinfrancois opened the pull request:
<!-- Please describe your changes here and list any open questions you might have. -->

This PR fixes a concurrency bug in `NamespacedHierarchicalStore.computeIfAbsent` where `defaultCreator` was executed while holding the internal map's bucket lock, causing:
1. Threads accessing different keys in the same hash bucket to block each other during parallel test execution
2. Potential deadlocks when `defaultCreator` accesses other keys that collide in the same hash bucket

## Root Cause

The previous implementation called `defaultCreator.apply(key)` inside `ConcurrentHashMap.compute()`, which holds the bucket lock. Any thread trying to access a *different* key in the same bucket was blocked until `defaultCreator` completed.

## Fix

The fix moves `defaultCreator` execution outside the lock:
- A placeholder (`MemoizingSupplier` or `DeferredSupplier`) is installed via the map operation
- The actual computation runs after the map operation completes
- Failed computations are cleaned up, and `get()` returns `null` (not the exception) during the transient window

**Note:** Two implementation approaches are provided as separate commits. See the comment below for details on each approach.

## Changes

### Implementation
- Reworked `computeIfAbsent` to execute `defaultCreator` outside the bucket lock
- Ensured `get()` doesn't observe transient exceptions from failed `computeIfAbsent` calls
- Preserved "one initialization per key" semantics

### Tests
- `simulateRaceConditionInComputeIfAbsentWithCollidingKeys`: Verifies correct initialization under contention with colliding keys
- `computeIfAbsentWithCollidingKeysDoesNotBlockConcurrentAccess`: Verifies no blocking between colliding keys
- `computeIfAbsentDoesNotDeadlockWithCollidingKeys`: Verifies no deadlock with colliding keys
- `getOrComputeIfAbsentDoesNotDeadlockWithCollidingKeys`: Same for deprecated method
- `getDoesNotSeeTransientExceptionFromComputeIfAbsent`: Verifies atomicity for transient failures
- `getConcurrentWithFailingComputeIfAbsentDoesNotSeeException`: Stress test for atomicity
- `computeIfAbsentOverridesParentNullValue`: Verifies parent/child store semantics
- Added `CollidingKey` helper to force hash collisions for testing

Fixes #5171

This should also fix the flakiness in AssertJ's `SoftAssertionsExtension_PER_CLASS_Concurrency_Test` (https://github.com/assertj/assertj/issues/1996).

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit-framework/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://docs.junit.org/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://docs.junit.org/snapshot/user-guide/) and [Release Notes](https://docs.junit.org/snapshot/user-guide/#release-notes)

Pankraz76 commented on junit-platform-engine/src/main/java/org/junit/platform/engine/support/store/NamespacedHierarchicalStore.java:
```diff
@@ -245,18 +246,21 @@ public <K, V> Object computeIfAbsent(N namespace, K key, Function<? super K, ? e
 		var result = StoredValue.evaluateIfNotNull(storedValue);
 		if (result == null) {
 			StoredValue newStoredValue = this.storedValues.compute(compositeKey, (__, oldStoredValue) -> {
```
```suggestion
			StoredValue newStoredValue = storedValues.compute(compositeKey, (__, oldStoredValue) -> {
```
the best we can do is to do nothing at all, or do whats close to that meaning delete.

everything we can delete without breaking seems obsolet, mostly imposing unwanted coupling burden.

Pankraz76 reviewed (commented):
2cts.

going fully functional separating the concerns (SoC/SRP).

martinfrancois commented on junit-platform-engine/src/main/java/org/junit/platform/engine/support/store/NamespacedHierarchicalStore.java:
This is a change in `getOrComputeIfAbsent` which is outside of the scope of this PR, but feel free to submit it as a separate PR as it's completely independent of my change.

martinfrancois commented:
You’re welcome, and thanks again, @Pankraz76, for taking another careful look.
Your suggestions are great, but as you pointed out, they mostly target existing code that could be cleaned up independently. I’d like to keep this PR focused on the bugfix so we do not drag out the review with additional refactoring.
Once this is merged, I am happy to follow up with a separate cleanup PR. I do not want to delay AssertJ being able to update to JUnit 6 again any longer than necessary 🙂

mpkorstanje commented on documentation/modules/ROOT/partials/release-notes/release-notes-6.0.2.adoc:
```diff
@@ -17,6 +17,13 @@ repository on GitHub.
+  block each other and temporarily see a missing or incorrectly initialized state
+  for values created via `computeIfAbsent`. The method now evaluates
+  `defaultCreator` outside the critical section using a memoizing supplier,
+  aligning its behavior with the deprecated `getOrComputeIfAbsent`.
```
Could you write this down more concisely?

mpkorstanje reviewed (commented):
I've given this a quick read through and left some comments to resolve open questions, but this is not a full review yet.

It seems that you've found and solved a different problem than described in https://github.com/junit-team/junit-framework/issues/5171. And you claim the solutions overlap.

Unfortunately the original problem is quite tricky and the description for this pull request incredibly verbose. We'll need some time to go through the details. You can help us process this by writing a much more concise PR description.

martinfrancois commented on platform-tests/src/test/java/org/junit/platform/engine/support/store/NamespacedHierarchicalStoreTests.java:
```diff
@@ -416,6 +419,115 @@ void simulateRaceConditionInComputeIfAbsent() throws Exception {
+		}
+
+		@Test
+		void computeIfAbsentCanDeadlockWithCollidingKeys() throws Exception {
```
Agreed. After the fix, this scenario should no longer deadlock. I renamed it to `computeIfAbsentDoesNotDeadlockWithCollidingKeys`

martinfrancois commented on junit-platform-engine/src/main/java/org/junit/platform/engine/support/store/NamespacedHierarchicalStore.java:
```diff
@@ -237,26 +237,30 @@ public void close() {
+				return requireNonNull(newStoredValue.evaluate());
+			}
+			catch (Throwable t) {
+				storedValues.remove(compositeKey, newStoredValue);
```
> And I think this invalidates any approach that tries to avoid execution of the defaultCreator outside the compute method.

Both implementations address this concern:

**Approach 1 (MemoizingSupplier):** Added `transientFailures` flag. When `true`, `get()` returns `null` instead of throwing during the transient window.

**Approach 2 (DeferredSupplier):** Uses type-based dispatch. `DeferredSupplier.get()` catches `ExecutionException` and returns `null`, while `getOrThrow()` rethrows for the original `computeIfAbsent` caller.

Both ensure:
- `defaultCreator` runs outside the bucket lock (avoiding blocking/deadlock)
- `get()` returns `null` during the transient window (not the exception)
- The original `computeIfAbsent` caller sees the exception and cleans up
- After cleanup, `get()` continues to return `null` (correct final state)

Tests added:
- `getDoesNotSeeTransientExceptionFromComputeIfAbsent`
- `getConcurrentWithFailingComputeIfAbsentDoesNotSeeException`

martinfrancois commented:
Thanks for the initial review @mpkorstanje!

I apologize for being a bit verbose on both the release notes and the description of the PR. I now shortened both of them, I hope I didn't shorten them too much - let me know if I should make any changes there.

I also addressed your comments by adding both tests to reproduce the issues you mentioned and also addressed them. I first implemented an approach but didn't like some aspects, so I tried another approach. Both approaches ended up working, and since I'm not sure which one you would prefer I decided to push both as separate commits, so I could later drop the one you don't like.

1. **Commit 1** (`deb65ab4`): Uses `MemoizingSupplier` with a `transientFailures` flag
   - Adds a boolean flag to distinguish between callers that should see exceptions vs. those that shouldn't
   - Uses `storedValues.compute()` with the supplier installed inside the lambda

2. **Commit 2** (`81d673b0`): Uses `DeferredSupplier` with `FutureTask`
   - Uses `DeferredSupplier` for `computeIfAbsent`, `MemoizingSupplier` for `getOrComputeIfAbsent`
   - Uses explicit CAS-retry loop (`for (;;)` with `putIfAbsent`/`replace`) instead of `compute()`
   - Leverages `FutureTask` which already provides "run once" semantics

Both approaches:
- Fix the bucket-lock contention issue (threads no longer block each other on colliding keys)
- Fix the deadlock scenario (first computation waiting for second no longer deadlocks)
- Ensure `get()` doesn't see transient exceptions from failed `computeIfAbsent`
- Pass all existing and new tests

**Trade-offs:**

| Aspect | Approach 1 (MemoizingSupplier) | Approach 2 (DeferredSupplier) |
|--------|-------------------------------|-------------------------------|
| Complexity | Less code, adds flag to existing class | More code, new class |
| Dispatch | Boolean flag (more implicit) | Type-based (cleaner, but more verbose) |
| Pattern | Uses `compute()` | Explicit CAS-retry loop (matches ConcurrentHashMap internals) |
| Dependencies | None | Uses `FutureTask` |

Please let me know which approach you prefer, and I'll drop the other commit.

mpkorstanje reviewed (commented):
Overall I think you're going in the right direction.

But I don't think the CAS loop is a good solution. It means that we're not taking advantage of the existing data structure.

Rather, I think it would be important to recognize and make explicit that there are 4 type of stored values. Absent (if nothing was stored yet), Constant (from put), Memoized (from getOrComputeIfAbsent) and Defered (from computeIfAbsent). These are containers that may have the following contents:

| Type         | Value    | Null | Exception |
|--------|--------|--------|--------|
| Absent | No |  No | No |
| Constant | Maybe |  Maybe | No |
| Memoized | Maybe | Maybe | Maybe |
| Defered | Maybe | No | Maybe |


These can be directly stored in the map and evaluated immediately afterwards. Then depending on retrieval function (get, getOrComputeIfAbsent, computeIfAbsent) different things happen.  This means that if the evaluation of a value failed, the value need need not be removed from the map. 


| Method | Type | Containing | Action |
| ----------- | ----------- | ----------- | ----------- |
| get |  Absent | - | return null |
| get |  Constant | value | return value |
| get |  Constant | null | return null |
| get |  Memoized | value | return value |
| get |  Memoized | null | return null |
| get |  Memoized | exception | trow exception |
| get |  Defered | value | return value |
| get |  Defered | exception | return null |
| getOrComputeIfAbsent |  Absent | - | compute new value |
| getOrComputeIfAbsent |  Constant | value | return value |
| getOrComputeIfAbsent |  Constant | null | return null |
| getOrComputeIfAbsent |  Memoized | value | return value |
| getOrComputeIfAbsent |  Memoized | null | return null |
| getOrComputeIfAbsent |  Memoized | exception | trow exception |
| getOrComputeIfAbsent |  Defered | value | return value |
| getOrComputeIfAbsent |  Defered | exception | return null |
| computeIfAbsent |  Absent | - | compute new value |
| computeIfAbsent |  Constant | value | return value |
| computeIfAbsent |  Constant | null | compute new value |
| computeIfAbsent |  Memoized | value | return value |
| computeIfAbsent |  Memoized | null | compute new  value |
| computeIfAbsent |  Memoized | exception | compute new  value |
| computeIfAbsent |  Defered | value | return value |
| computeIfAbsent |  Defered | exception | compute new  value |


You have addressed some this implicitly already but I think it would be good to make this explicit.

For example one of the cases that was overlooked in the current situation happens when a call to `getOrComputeIfAbsent` puts a value in the store but `computeIfAbsent` wins the race to evaluates it. This would allow an exception to be thrown from `computeIfAbsent` that did not originate from its creator function. I think the solution would be to have both `computeIfAbsent` and `getOrComputeIfAbsent` use the `DeferredSupplier`.

Finally, I think the current solution is also very dependent on testing race conditions. This make for rather verbose and potentially flakey tests. I think we can avoid most of that by letting the concurrent hash-map do the heavy lifting and then figure out what to do with it state afterwards. 

Anyway, I'm going to have a look tomorrow and see if I can get all of those concepts into a pull request.

mpkorstanje commented:
Cheers! It was a bit easier to start writing tests on a new branch and copy the relevant bits over. I've given you co-author credits in the git commits!

Superseded by #5231.
